companion.py

```diff
@@ -463,27 +520,119 @@ def base64_url_encode(self, text: bytes) -> str:
+        play_sound: bool = False, auto_update: bool = False
+    ):
+        super().__init__(
+            query_time=query_time, tk_response_event=tk_response_event, retrying=retrying,
+            play_sound=play_sound, auto_update=auto_update
+        )
+        self.endpoint: str = endpoint  # The CAPI query to perform.
+
+
+class EDMCCAPIResponse(EDMCCAPIReturn):
+    """Encapsulates a response from CAPI quer(y|ies)."""
+
+    def __init__(
+            self, capi_data: CAPIData,
+            query_time: int, retrying: bool = False, play_sound: bool = False, auto_update: bool = False
+    ):
+        super().__init__(query_time=query_time, retrying=retrying, play_sound=play_sound, auto_update=auto_update)
+        self.capi_data: CAPIData = capi_data  # Frontier CAPI response, possibly augmented (station query)
+
+
+class CAPIFailedRequest(EDMCCAPIReturn):
+    """CAPI failed query error class."""
+
+    def __init__(
+            self, message: str,
+            query_time: int, retrying: bool = False, play_sound: bool = False, auto_update: bool = False,
+            exception=None
+    ):
+        super().__init__(query_time=query_time, retrying=retrying, play_sound=play_sound, auto_update=auto_update)
+        self.message: str = message  # User-friendly reason for failure.
+        self.exception: int = exception  # Exception that recipient should raise.
+
+
 class Session(object):
-    """Methods for handling an oAuth2 session."""
+    """Methods for handling Frontier Auth and CAPI queries."""
 
     STATE_INIT, STATE_AUTH, STATE_OK = list(range(3))
 
+    FRONTIER_CAPI_PATH_PROFILE = '/profile'
+    FRONTIER_CAPI_PATH_MARKET = '/market'
+    FRONTIER_CAPI_PATH_SHIPYARD = '/shipyard'
+    # This is a dummy value, to signal to Session.capi_query_worker that we
+    # the 'station' triplet of queries.
+    _CAPI_PATH_STATION = '_edmc_station'
+
     def __init__(self) -> None:
         self.state = Session.STATE_INIT
         self.server: Optional[str] = None
         self.credentials: Optional[Dict[str, Any]] = None
         self.session: Optional[requests.Session] = None
         self.auth: Optional[Auth] = None
         self.retrying = False  # Avoid infinite loop when successful auth / unsuccessful query
+        self.tk_master: Optional[tk.Tk] = None
+
+        self.capi_raw_data = CAPIDataRaw()
+        logger.debug('Starting CAPI queries thread...')
+        self.capi_response_queue: Queue
+        self.capi_query_queue: Queue = Queue()
+        self.capi_query_thread = threading.Thread(
+            target=self.capi_query_worker,
+            daemon=True,
+            name='CAPI worker'
+        )
+        self.capi_query_thread.start()
+        logger.debug('Done')
+
+    def set_capi_response_queue(self, capi_response_queue: Queue) -> None:
+        """Set a reference to the CAPI response data queue."""
+        self.capi_response_queue = capi_response_queue
```

Comment:
does this need to be a setter?
if so, would a property not work better?

Reply:
Actually I'll rework this to create the queue in Session now.  This is mostly an artifact of the order I started changing the code.  The callers will have `companion.session` to reference it.